File: stringToInteger/stringToInteger.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int myAtoi(string str) {
        long long res = 0;
        int i = 0;
        int sign = 1;
        while(str[i] == ' ') i++;
        if(str[i] == '+' || str[i] == '-') {
            if(str[i] == '-') sign = -1;
            i++;
        }
        while(str[i] >= '0' && str[i] <= '9') {
            res *= 10;
            res += (int) (str[i] - '0');
            i++;
            if (res * sign > INT_MAX)
                return INT_MAX;
            else if(res * sign < INT_MIN)
                return INT_MIN; 
        }
        return res * sign;  
    }
};

int main()
{
    Solution s;
    string ss = "-1123111111111111111";
    //int a = atoi(ss.c_str());
  
    cout << s.myAtoi(ss) << endl;
    return 0;
}

File: 8.stringToInteger/stringToInteger.java
class Solution {
    public int myAtoi(String s) {
        int length = s.length();
        int i = 0;
        
        // skip whitespace
        while (i < length && s.charAt(i) == ' ') {
            i++;
        }

        // singedness
        int sign = 1;
        if (i < length && (s.charAt(i) == '+' || s.charAt(i) == '-')) {
            if (s.charAt(i) == '-') sign = -1;
            i++;
        }

        long res = 0;

        // conversion
        while (i < length) {
            char c = s.charAt(i);
            if (c < '0' || c > '9') break;
            int digit = c - '0';
            res = res * 10 + digit;
            long signedRes = res * sign;
            if (signedRes > Integer.MAX_VALUE) {
                return Integer.MAX_VALUE;
            } else if (signedRes < Integer.MIN_VALUE) {
                return Integer.MIN_VALUE;
            }
            i++;
        }
        return (int) (sign * res);
    }
}